### PR TITLE
improve integration test and pipeline

### PR DESCRIPTION
**Pull request by rafaelleinio**
# Description
A collection of improvements focused on the integration test and the pipeline definition

# What's new? 
- :bug: fix all current warnings in tests
- 🧪 increase coverage to 100%
- 🔧 improve the organization of the integration test
- 🔧 improve the organization of the pipeline definition
- :sparkles: add new badges to README

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: scripts/deaths_per_year_run.py [hunk 1]
```diff
@@ -1,9 +1,10 @@
-from spark_pipeline.deaths_per_year import extract, load, transform
+from spark_pipeline import deaths_per_year_pipeline
 
 
 def main() -> None:
     """Main."""
-    load(transform(extract()))
+    print(">>> Pipeline execution starting...")
+    deaths_per_year_pipeline.run()
```
Comment (rafaelleinio) [inline scripts/deaths_per_year_run.py:7]: cleaner right?

File: spark_pipeline/base/read_write.py [hunk 2]
```diff
@@ -27,19 +25,17 @@ def read_csv(
...
+        mode: write mode. E.g. 'append' or 'overwrite'
+
     """
-    warnings.warn(
-        "Warning: This function does not use distributed operation,"
-        "\n"
-        "so it might not be a good solution to bigger datasets."
-    )
-    df.coalesce(1).write.parquet(path=path, mode="overwrite")
+    df.coalesce(output_partitions).write.mode(mode).parquet(path=path, mode="overwrite")
```
Comment (rafaelleinio) [inline spark_pipeline/base/read_write.py:41]: I think this warning was not making too much sense for the context... in any way, I've added parameters for the user to tune this writing operation as needed

File: tests/integration/spark_pipeline/test_deaths_per_year_pipeline/input.csv [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Id,Name,Death_year
+Q1,Peter,2000
+Q2,Bob,2000
+Q3,John,2000
+Q4,Mary,2001
+Q5,Susan,2002
```
Comment (rafaelleinio) [inline tests/integration/spark_pipeline/test_deaths_per_year_pipeline/input.csv:6]: I've simplified test case, as your pipeline only read these columns

File: tests/integration/spark_pipeline/test_deaths_per_year_pipeline/expected_output.csv [hunk 4]
```diff
@@ -0,0 +1,4 @@
+total_deaths,death_year
+3,2000
+1,2001
+1,2002
```
Comment (rafaelleinio) [inline tests/integration/spark_pipeline/test_deaths_per_year_pipeline/expected_output.csv:4]: now your target output is much simple to understand as well

File: tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        input_path=f"{CURRENT_PATH}/input.csv"
+    )
+
+    # act
+    deaths_per_year_pipeline.run(pipeline)
+    output_df: DataFrame = mock_write_jdbc.call_args.kwargs["df"]
+
+    # assert
+    assert sorted(output_df.collect()) == sorted(expected_df.collect())
+    assert mock_write_jdbc.call_args.kwargs["mode"] == "overwrite"
```
Comment (rafaelleinio) [inline tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py:38]: I tried to improve the organization of this integration test. I think is much cleaner now right? I always like to have these `arrange`, `act`, and `assert` stages well organized and defined, it helps in the overall readability of the test.

File: tests/unit/spark_pipeline/base/test_read_write.py [hunk 6]
```diff
@@ -32,6 +32,6 @@ def test_read_csv() -> None:
 def test_write_csv(mock_load: Mock) -> None:
     input = [{"test": 1, "test_2": 2}, {"test": 1, "test_2": 2}]
     spark = SparkSession.builder.getOrCreate()
-    df1 = spark.createDataFrame(input)
+    df1 = spark.createDataFrame(input, schema="test int, test_2 int")
```
Comment (rafaelleinio) [inline tests/unit/spark_pipeline/base/test_read_write.py:35]: this fix a warning in the test

File: spark_pipeline/deaths_per_year_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            df (DataFrame): Aggregated DataFrame
+        """
+        read_write.write_csv(df=df, path=self.output_path, mode="overwrite")
+
+
+def run(pipeline: DeathsPerYearPipeline = DeathsPerYearPipeline()) -> None:
+    """Run DeathsPerYearPipeline pipeline."""
+    extract_df = pipeline.extract()
+    transform_df = pipeline.transform(extract_df)
+    return pipeline.load(transform_df)
```
Comment (rafaelleinio) [inline spark_pipeline/deaths_per_year_pipeline.py:67]: I've just organized your functions into a class to define the pipeline. I've added attributes for the user to be able to parametrize some configurations like the input and output paths. Besides, in the write function, you can see that I've replaced the manual file glob/remove logic for Spark's `overwrite` write mode, which is doing the same thing but in a simpler way.

**Review by rafaelleinio — COMMENTED** [review]
just a few more explanations about my changes:

File: tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py (hunk 5 above)
Comment (joao-victor-campos) [reply tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py:38]: What does this act stage do exactly? Is it creating a mock object?

Comment (rafaelleinio) [reply tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py]: run the pipeline, and capture the final df that is passed to the `write_csv` service

Comment (rafaelleinio) [reply tests/integration/spark_pipeline/test_deaths_per_year_pipeline/test_death_per_year.py]: sorry, I have made a mistake while naming this mock object, here the fix: https://github.com/joao-victor-campos/pyspark-data-aggregation-pipeline/pull/14/commits/aa3a7928db47e34cb7acd04719ed8a0caf0f433d

**Review by joao-victor-campos — APPROVED** [review]
All changes were a great addition to simplify and improve the code, ty 😎

**Review by NicholasBaraldi — APPROVED** [review]
Thank you! Is nice to see a more senior way to handle this kind of problem.

**Pull request merged**